pkg/autoscaler/autoscaler.go

```diff
@@ -291,6 +294,106 @@ var _ = Describe("Autoscaler should", framework.LabelAutoscaler, Serial, func()
 			}, framework.WaitLong, pollingInterval).Should(BeTrue(), "MachineSet %s failed to scale in to 0 replicas", machineSet.GetName())
 		})
 
+		// Machines required for test: 1
+		// We should need 1 only machineset. In case of failure, we might get more than 1 machineset.
+		// Reason: This test checks that the autoscaler is able to scale from zero when a workload requires specific architecture in the node affinity fields.
+		// Moreover, this test gives a better signal when multiple architectures are available in the cluster or the cluster is not amd64,
+		// as the workload is set to be scheduled on an architecture different from amd64.
+		It("It scales from/to zero a machine set with the architecture requested by the workload", func() {
+			// Only run in platforms which support arch-aware autoscaling from/to zero.
+			clusterInfra, err := framework.GetInfrastructure(ctx, client)
+			Expect(err).NotTo(HaveOccurred(), "Failed to get cluster infrastructure object")
+
+			platform := clusterInfra.Status.PlatformStatus.Type
+			switch platform {
+			case configv1.AWSPlatformType, configv1.AzurePlatformType:
+				klog.Infof("Platform is %v", platform)
+			default:
+				Skip(fmt.Sprintf("Platform %v does not support arch-aware autoscaling from/to zero, skipping.", platform))
+			}
+
+			By("Creating a new MachineSet with 0 replicas for each machineset of a different architecture found in the cluster")
+			expectedReplicas := int32(1)
+			machineSetParamsList := framework.BuildPerArchMachineSetParamsList(ctx, client, 0)
+			machineSets := make([]*machinev1.MachineSet, 0)
+			targetedNodeLabel := fmt.Sprintf("%v-scale-from-zero", autoscalerWorkerNodeRoleLabel)
+			var expectedScaledMachineSet *machinev1.MachineSet
+			var workloadArch string
+
+			for _, machineSetParams := range machineSetParamsList {
+				machineSetParams.Labels[targetedNodeLabel] = ""
+				By(fmt.Sprintf("Deploying the MachineSet with 0 replicas (architecture: %s)",
+					machineSetParams.Labels[framework.ArchLabel]))
+				machineSet, err := framework.CreateMachineSet(client, machineSetParams)
+				Expect(err).ToNot(HaveOccurred(), "Failed to create MachineSet with 0 replicas")
+				machineSets = append(machineSets, machineSet)
+				cleanupObjects[machineSet.GetName()] = machineSet
+
+				framework.WaitForMachineSet(ctx, client, machineSet.GetName())
+
+				By(fmt.Sprintf("Creating a MachineAutoscaler backed by MachineSet %s/%s - min:%v, max:%v",
+					machineSet.GetNamespace(), machineSet.GetName(), 0, expectedReplicas))
+				asr := machineAutoscalerResource(machineSet, 0, expectedReplicas)
+				Expect(client.Create(ctx, asr)).Should(Succeed(), fmt.Sprintf("Failed to create MachineAutoscaler with min 0/max %v replicas", expectedReplicas))
+				cleanupObjects[asr.GetName()] = asr
+				arch := machineSetParams.Labels[framework.ArchLabel]
+				if arch != framework.Amd64 || workloadArch == "" {
```

Comment:
just to confirm, given the logic around `workloadArch` this clause will always be entered on the first loop of the `for`, is that accurate?

Reply:
Yes, it is.

The var is set to the empty string first so that the `or` operation results true during the first iteration and set both the workloadArch and expectedScaledOutMachineSet to correct and paired values.

The other iterations will try to choose a non amd64 machine set for the test case. The last non amd64 machine set is the expected machine set to scale out.